R/decisiontreeutils.R

```diff
@@ -0,0 +1,431 @@
+  # inheritance for printing and plotting purposes).
+  if (is_XDF) {
+    the_model_rpart <- rxAddInheritance(the_model)
+    printcp(the_model_rpart)
+    out <- capture.output(printcp(the_model_rpart))
+    the_model$xlevels <- do.call(match.fun("xdfLevels"), list(paste0("~ ", paste(names_x_vars, collapse = " + ")), xdf_path))
+    if (is.factor(the_target)) {
+      target_info <- do.call(match.fun("rxSummary"), list(paste0("~ ", name_y_var), data = xdf.path))[["categorical"]]
+      if(length(target_info) == 1) {
+        the_model$yinfo <- list(levels = as.character(target_info[[1]][,1]), counts = target_info[[1]][,2])
+      }
+    }
+  } else {
+    printcp(the_model) # Pruning Table
+    out <- capture.output(printcp(the_model))
+  }
+
+  out <- out[out != ""]
+  print(out)
+  out <- out[2:length(out)]
+  # Find the index of the end of the call
+  for (i in 1:length(out)) {
+    if (grepl("Variables", out[i])) {
+      end_call <- i - 1
+      break
+    }
+  }
+  call1 <- out[1:end_call]
+  the_call = gsub("\\s\\s", "", paste(call1, collapse=""))
+  # Figure out where the summary information ends and the actual pruning table
+  # starts. The header for the pruning table is removed, and will instead show-up
+  # in the Alteryx reporting tool
+  for (i in 1:length(out)) {
+    if (grepl("\\sCP\\s", out[i])) {
+      strt_table <- i + 1
+      break
+    }
+  }
+  model_sum <- length((end_call + 1):(strt_table - 2))
+  prune_tbl1 <- out[strt_table:length(out)]
+  out <- c(the_call, out[(end_call + 1):(strt_table - 2)])
+  rpart_out <- data.frame(grp = c("Call", rep("Model_Sum", model_sum)), out = out)
+  rpart_out$grp <- as.character(rpart_out$grp)
+  rpart_out$out <- as.character(out)
+  # Pipe delimit the pruning table and then rbind it to the output
+  prune_tbl <- NULL
+  for (i in 1:length(prune_tbl1)) {
+    a_row <- unlist(strsplit(prune_tbl1[i], "\\s"))
+    a_row <- a_row[a_row != ""]
+    prune_tbl <- c(prune_tbl, paste(a_row[1], a_row[2], a_row[3], a_row[4], a_row[5], a_row[6], sep="|"))
+  }
+  pt_df <- data.frame(grp = rep("Prune", length(prune_tbl)), out = prune_tbl)
+  pt_df$grp <- as.character(pt_df$grp)
+  pt_df$out <- as.character(pt_df$out)
+  rpart_out <- rbind(rpart_out, pt_df)
+
+  # The leaf summary
+  if (is_XDF) {
+    print(the_model_rpart) # Tree Leaf Summary
+    leaves <- capture.output(the_model_rpart)
+  } else {
+    print(the_model) # Tree Leaf Summary
+    leaves <- capture.output(the_model)
+  }
+  leaves_num <- 1:length(leaves)
+  start_leaves <- leaves_num[substr(leaves, 1, 5) == "node)"]
+  leaves <- leaves[start_leaves:length(leaves)]
+  leaves <- gsub(">", "&gt;", leaves)
+  leaves <- gsub("<", "&lt;", leaves)
+  leaves <- gsub("\\s", "<nbsp/>", leaves) # Order matters
+  leaves_df <- data.frame(grp = rep("Leaves", length(leaves)), out = leaves)
+  leaves_df$grp <- as.character(leaves_df$grp)
+  leaves_df$out <- as.character(leaves_df$out)
+
```

Comment:
The readability of this code can be improved by using the piping syntax provided by the `magrittr` package. It is akin to chaining in `javascript` and IMHO makes the code read more like poetry.

```r
  capture.output(model) %>%
    extract(grep("^node", .):length(.)) %>%
    gsub(">", "&gt;") %>%
    gsub("<", "&lt;") %>%
    gsub("\\s", "<nbsp/>", .) %>%
    data.frame(grp = "Leaves", out = ., stringsAsFactors = FALSE)
```

In addition, I would throw this into a function named `extractLeaves` so that we can test the output if necessary.